File: packages/ios/Sources/CrumbUI/CrumbUploadCoordinator.swift
import CrumbCore
#if canImport(UIKit) && canImport(Network)
import Network
import UIKit

@MainActor
final class CrumbUploadCoordinator {
    static let shared = CrumbUploadCoordinator()

    private let monitorQueue = DispatchQueue(label: "dev.crumb.upload-connectivity")
    private var worker: CrumbReportUploadWorker?
    private var uploadTask: Task<Void, Never>?
    private var retryTask: Task<Void, Never>?
    private var pathMonitor: NWPathMonitor?
    private var retryStep = 0
    private var isForeground = false

    private init() {}

    func resume() {
        isForeground = true
        wake()
    }

    func pause() {
        isForeground = false
        retryTask?.cancel()
        retryTask = nil
        uploadTask?.cancel()
        stopConnectivityMonitoring()
    }

    func reportDidQueue() {
        wake()
    }

    private func wake() {
        guard isForeground else { return }
        if worker == nil {
            guard let settings = try? Crumb.uploadSettings() else { return }
            worker = CrumbReportUploadWorker(settings: settings)
        }
        guard let worker else { return }

        startConnectivityMonitoring()
        guard uploadTask == nil else { return }
        retryTask?.cancel()
        retryTask = nil
        uploadTask = Task { [weak self] in
            let result = await worker.runPass()
            guard let self else { return }
            self.uploadTask = nil
            self.handle(result)
        }
    }

    private func handle(_ result: CrumbUploadPassResult) {
        if result.remainingReportCount == 0 {
            retryStep = 0
            stopConnectivityMonitoring()
            return
        }
        guard isForeground else { return }
        if result.wasCancelled {
            wake()
        } else if result.shouldRetry {
            scheduleRetry()
        } else {
            // Permanent failures remain durable, but do not consume battery in a retry loop.
            stopConnectivityMonitoring()
        }
    }

    private func scheduleRetry() {
        let delays: [UInt64] = [1, 2, 4, 8, 16, 32, 60]
        let seconds = delays[min(retryStep, delays.count - 1)]
        retryStep = min(retryStep + 1, delays.count - 1)
        retryTask?.cancel()
        retryTask = Task { [weak self] in
            do {
                try await Task.sleep(nanoseconds: seconds * 1_000_000_000)
            } catch {
                return
            }
            guard let self else { return }
            self.retryTask = nil
            self.wake()
        }
    }

    private func startConnectivityMonitoring() {
        guard pathMonitor == nil else { return }
        let monitor = NWPathMonitor()
        monitor.pathUpdateHandler = { [weak self] path in
            guard path.status == .satisfied else { return }
            Task { @MainActor [weak self] in
                guard let self, self.isForeground else { return }
                self.retryStep = 0
                self.retryTask?.cancel()
                self.retryTask = nil
                self.wake()
            }
        }
        pathMonitor = monitor
        monitor.start(queue: monitorQueue)
    }

    private func stopConnectivityMonitoring() {
        pathMonitor?.cancel()
        pathMonitor = nil
    }
}
#endif

File: packages/android/crumb-ui/src/main/kotlin/dev/crumb/ui/CrumbUploadCoordinator.kt
package dev.crumb.ui

import android.app.Application
import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.os.Handler
import android.os.Looper
import dev.crumb.core.Crumb
import dev.crumb.core.CrumbReportQueue
import dev.crumb.core.CrumbReportUploadWorker
import dev.crumb.core.CrumbUploadPassResult

internal object CrumbUploadCoordinator {
    private val mainHandler = Handler(Looper.getMainLooper())
    private val retryDelaysMillis = longArrayOf(1_000, 2_000, 4_000, 8_000, 16_000, 32_000, 60_000)

    private var application: Application? = null
    private var worker: CrumbReportUploadWorker? = null
    private var running = false
    private var foreground = false
    private var retryStep = 0
    private var retryRunnable: Runnable? = null
    private var networkCallback: ConnectivityManager.NetworkCallback? = null

    fun install(application: Application) {
        if (this.application != null) return
        val settings = runCatching { Crumb.uploadSettings() }.getOrNull() ?: return
        this.application = application
        worker = CrumbReportUploadWorker.create(CrumbReportQueue.from(application), settings)
    }

    fun resume(application: Application) {
        install(application)
        foreground = true
        wake()
    }

    fun pause() {
        foreground = false
        retryRunnable?.let(mainHandler::removeCallbacks)
        retryRunnable = null
        worker?.cancel()
        stopConnectivityMonitoring()
    }

    fun reportDidQueue() {
        wake()
    }

    private fun wake() {
        val activeWorker = worker ?: return
        if (!foreground || running) return
        startConnectivityMonitoring()
        retryRunnable?.let(mainHandler::removeCallbacks)
        retryRunnable = null
        running = true
        activeWorker.prepareForPass()
        Thread({
            val result = activeWorker.runPass()
            mainHandler.post {
                running = false
                handle(result)
            }
        }, "Crumb report upload").start()
    }

    private fun handle(result: CrumbUploadPassResult) {
        if (result.remainingReportCount == 0) {
            retryStep = 0
            stopConnectivityMonitoring()
            return
        }
        if (!foreground) return
        when {
            result.wasCancelled -> wake()
            result.shouldRetry -> scheduleRetry()
            else -> stopConnectivityMonitoring()
        }
    }

    private fun scheduleRetry() {
        val delay = retryDelaysMillis[retryStep.coerceAtMost(retryDelaysMillis.lastIndex)]
        retryStep = (retryStep + 1).coerceAtMost(retryDelaysMillis.lastIndex)
        retryRunnable?.let(mainHandler::removeCallbacks)
        val runnable = Runnable {
            retryRunnable = null
            wake()
        }
        retryRunnable = runnable
        mainHandler.postDelayed(runnable, delay)
    }

    private fun startConnectivityMonitoring() {
        if (networkCallback != null) return
        val app = application ?: return
        val manager = app.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            ?: return
        val callback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                mainHandler.post {
                    if (!foreground) return@post
                    retryStep = 0
                    retryRunnable?.let(mainHandler::removeCallbacks)
                    retryRunnable = null
                    wake()
                }
            }
        }
        try {
            manager.registerDefaultNetworkCallback(callback)
            networkCallback = callback
        } catch (_: SecurityException) {
            // Timed retry remains available if a host manifest strips the merged permission.
        } catch (_: RuntimeException) {
            // Local submission and durable retry do not depend on connectivity observation.
        }
    }

    private fun stopConnectivityMonitoring() {
        val callback = networkCallback ?: return
        val app = application
        val manager = app?.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
        runCatching { manager?.unregisterNetworkCallback(callback) }
        networkCallback = null
    }
}
